From c3c7fb30af165743ac4c6418af3ec85324971bbe Mon Sep 17 00:00:00 2001
From: Alexey Dobriyan <adobriyan@openvz.org>
Date: Fri, 7 Mar 2008 17:56:41 +0300
Subject: [PATCH 15/48] NETFILTER: fixup net.netfilter sysctl data for VE0 too

They were pointing to usual mainline variables, not VE0's
and, thus, e.g, net.netfilter.nf_conntrack_count was always
reporting 0.
---
 net/netfilter/nf_conntrack_standalone.c |   23 ++++++++++-------------
 1 files changed, 10 insertions(+), 13 deletions(-)

diff --git a/net/netfilter/nf_conntrack_standalone.c b/net/netfilter/nf_conntrack_standalone.c
index fb5e4c9..1f33841 100644
--- a/net/netfilter/nf_conntrack_standalone.c
+++ b/net/netfilter/nf_conntrack_standalone.c
@@ -453,19 +453,15 @@ EXPORT_SYMBOL_GPL(nf_ct_log_invalid);
 #if defined(CONFIG_SYSCTL) && defined(CONFIG_VE_IPTABLES)
 static int nf_conntrack_init_ve_sysctl(struct ve_struct *ve)
 {
-	if (ve_is_super(ve)) {
-		ve_nf_ct_net_table = nf_ct_net_table;
-		ve_nf_ct_netfilter_table = nf_ct_netfilter_table;
-		ve_nf_ct_sysctl_table = nf_ct_sysctl_table;
-		ve_nf_ct_sysctl_header = register_sysctl_table(ve_nf_ct_net_table);
-		if (!ve_nf_ct_sysctl_header)
-			return -ENOMEM;
-		return 0;
-	}
+	ve_nf_ct_net_table = nf_ct_net_table;
+	ve_nf_ct_netfilter_table = nf_ct_netfilter_table;
+	ve_nf_ct_sysctl_table = nf_ct_sysctl_table;
 
-	ve_nf_ct_net_table = clone_sysctl_template(nf_ct_net_table);
-	if (ve_nf_ct_net_table == NULL)
-		goto out;
+	if (!ve_is_super(ve)) {
+		ve_nf_ct_net_table = clone_sysctl_template(nf_ct_net_table);
+		if (ve_nf_ct_net_table == NULL)
+			goto out;
+	}
 
 	ve_nf_ct_netfilter_table = ve_nf_ct_net_table[0].child;
 	ve_nf_ct_netfilter_table[1].data = &ve_nf_conntrack_max;
@@ -483,7 +479,8 @@ static int nf_conntrack_init_ve_sysctl(struct ve_struct *ve)
 	return 0;
 
 out_unclone:
-	free_sysctl_clone(ve_nf_ct_net_table);
+	if (!ve_is_super(ve))
+		free_sysctl_clone(ve_nf_ct_net_table);
 out:
 	return -ENOMEM;
 }
-- 
1.5.4.3

